src/main/kotlin/org/rust/ide/refactoring/changeSignature/impl.kt

```diff
@@ -0,0 +1,302 @@
+        parametersCopy.selfParameter,
+        parametersCopy.valueParameterList,
+        config
+    ) { factory.createValueParameter(it.patText, it.typeReference, reference = false) }
+}
+
+private fun renameParameter(factory: RsPsiFactory, parameter: RsValueParameter, newName: String) {
+    val identifier = factory.createIdentifier(newName)
+
+    val binding = parameter.pat?.findBinding()
+    if (binding != null) {
+        val usages = RenameUtil.findUsages(binding, newName, false, false, emptyMap())
+        for (info in usages) {
+            RenameUtil.rename(info, newName)
+        }
+    }
+
+    binding?.identifier?.replace(identifier)
+}
+
+private fun fixParametersOrder(
+    factory: RsPsiFactory,
+    /** [RsValueArgumentList] or [RsValueParameterList] */
+    parameters: RsElement,
+    parametersCopy: RsElement,
+    parameterSelf: RsElement?,
+    parametersList: List<RsElement>,
+    config: RsChangeFunctionSignatureConfig,
+    createNewParameter: (Parameter) -> PsiElement?
+) {
+    if (parametersList.size != config.originalParameters.size) return
+    val descriptors = config.parameters
+
+    // remove old parameters
+    val lparen = parameters.firstChild
+    val rparen = parameters.lastChild
+    if (lparen.nextSibling != rparen) {
+        parameters.deleteChildRange(lparen.nextSibling, rparen.prevSibling)
+    }
+    if (descriptors.isEmpty() && parameterSelf == null) return
+
+    // collect parameters psi in right order
+    val isMultiline = parametersCopy.textContains('\n')
+    val selfGroup = parameterSelf?.collectSurroundingWhiteSpaceAndComments()
+    val groupsWithoutSelf = descriptors.map { descriptor ->
+        if (descriptor.index == ParameterInfo.NEW_PARAMETER) {
+            val newline = if (isMultiline) factory.createNewline() else null
+            val psi = createNewParameter(descriptor)
+            listOfNotNull(newline, psi)
+        } else {
+            val psi = parametersList[descriptor.index]
+            psi.collectSurroundingWhiteSpaceAndComments()
+        }
+    }
+    val groups = listOfNotNull(selfGroup) + groupsWithoutSelf
+
+    // add parameters one by one
+    val rspace = parametersCopy.lastChild.prevSibling as? PsiWhiteSpace
+    val isSingleParameter = groups.size == 1 && descriptors.size < config.originalParameters.size
+    val hasTrailingComma = parametersList.lastOrNull()?.getNextNonCommentSibling()?.elementType == COMMA
+    for (group in groups) {
+        if (group !== groups.first()) {
+            parameters.addBefore(factory.createComma(), rparen)
+        }
+        for (element in group) {
+            if (element === rspace) continue  // will be added after loop
+            parameters.addBefore(element, rparen)
+        }
+    }
+
+    // Keep trailing comma if the resulting parameter list will have at least two parameters on multiple lines
+    if (hasTrailingComma &&
+        !isSingleParameter &&
+        (descriptors.isNotEmpty() || parameterSelf != null) &&
```

Comment:
```suggestion
```
this part is always true because of `if (descriptors.isEmpty() && parameterSelf == null) return` line above. Also maybe move this `if` inside else branch of `if (isSingleParameter) {`

Reply:
Sorry for this mess :D I originally used a different condition at the beginning of the function and forgot that it covers this scenario. So in the end the only `diff` is `if (hasTrailingComma)` -> `if (hasTrailingComma && isMultiline)` :man_facepalming: